File: app/src/main/java/com/example/weatherchallengemvvmkotlin/WeatherMVVMKotlinApplication.java
package com.example.weatherchallengemvvmkotlin;

import android.app.Activity;
import android.app.Application;
import android.os.Bundle;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;

import com.example.weatherchallengemvvmkotlin.injection.ApplicationComponent;
import com.example.weatherchallengemvvmkotlin.injection.ApplicationModule;
import com.example.weatherchallengemvvmkotlin.injection.ComponentObject;
import com.example.weatherchallengemvvmkotlin.injection.DaggerApplicationComponent;

import timber.log.Timber;


public class WeatherMVVMKotlinApplication extends Application implements Application.ActivityLifecycleCallbacks, ComponentObject<ApplicationComponent> {

    private static WeatherMVVMKotlinApplication instance;
    private ApplicationComponent applicationComponent;

    public static WeatherMVVMKotlinApplication getInstance() {
        return instance;
    }

    @Override
    public void onCreate() {
        super.onCreate();
        instance = this;
        Timber.plant(new Timber.DebugTree());
        registerActivityLifecycleCallbacks(this);
        ApplicationModule appModule = new ApplicationModule(this);
        applicationComponent = DaggerApplicationComponent.builder().applicationModule(appModule).build();
        appModule.setActivityComponent(applicationComponent);
        getAppComponent().inject(this);
    }

    public ApplicationComponent getAppComponent() {
        return applicationComponent;
    }

    @Override
    public void onActivityCreated(@NonNull Activity activity, @Nullable Bundle bundle) {

    }

    @Override
    public void onActivityStarted(@NonNull Activity activity) {

    }

    @Override
    public void onActivityResumed(@NonNull Activity activity) {

    }

    @Override
    public void onActivityPaused(@NonNull Activity activity) {

    }

    @Override
    public void onActivityStopped(@NonNull Activity activity) {

    }

    @Override
    public void onActivitySaveInstanceState(@NonNull Activity activity, @NonNull Bundle bundle) {

    }

    @Override
    public void onActivityDestroyed(@NonNull Activity activity) {

    }

    @Override
    public ApplicationComponent getComponent() {
        return null;
    }
}

File: app/src/main/java/com/example/weatherchallengemvvmkotlin/WeatherMVVMKotlinApplication.kt
package com.example.weatherchallengemvvmkotlin

import android.app.Activity
import android.app.Application
import android.os.Bundle
import com.example.weatherchallengemvvmkotlin.injection.ApplicationComponent
import com.example.weatherchallengemvvmkotlin.injection.ApplicationModule
import com.example.weatherchallengemvvmkotlin.injection.ComponentObject
import timber.log.Timber

class WeatherMVVMKotlinApplication: Application(), Application.ActivityLifecycleCallbacks, ComponentObject<ApplicationComponent?> {
    var appComponent: ApplicationComponent? = null
    private set

    companion object {
        lateinit var instance: WeatherMVVMKotlinApplication
    }

    override fun onCreate() {
        super.onCreate()
        instance = this
        Timber.plant(Timber.DebugTree())
        registerActivityLifecycleCallbacks(this)
        val appModule = ApplicationModule(this)
        appComponent = DaggerApplicationComponent.builder().applicationModule(appModule).build()
        appComponent?.inject(this)
        appModule.setApplicationComponent(appComponent!!)
    }

    override fun onActivityCreated(activity: Activity, bundle: Bundle?) {}
    override fun onActivityStarted(activity: Activity) {}
    override fun onActivityResumed(activity: Activity) {}
    override fun onActivityPaused(activity: Activity) {}
    override fun onActivityStopped(activity: Activity) {}
    override fun onActivitySaveInstanceState(activity: Activity, bundle: Bundle) {}
    override fun onActivityDestroyed(activity: Activity) {}
    override val component: ApplicationComponent?
        get() = null
}